src/Dataset.ts

```diff
@@ -508,18 +534,27 @@ export default class Dataset {
     return new Asset(this, await Asset["uploadAsset"]({ fileOrPath, assetName, dataset: this }));
   }
 
-  async addService(type: Models.ServiceTypeV1 | Models.ServiceTypeV2, name: string, reasoner?: Models.JenaReasoners) {
+  public async addService(name: string, opts?: NewService) {
     return new Service({
       app: this._app,
-      datasetPath: await this._getDatasetPath(),
-      datasetNameWithOwner: await this._getDatasetNameWithOwner(),
+      dataset: this,
       name,
-      type,
-      reasoner,
+      type: opts ? opts.type : "virtuoso",
+      reasoner: opts?.config?.reasoner ? opts?.config?.reasoner : undefined,
     }).create();
   }
 
-  async addPrefixes(newPrefixes: { [key: string]: string }) {
+  public async ensureService<T extends Dataset>(this: T, name: string, args?: NewService) {
+    try {
+      const foundService = await this.getService(name);
+      return foundService;
+    } catch (e: any) {
+      if (e.statusCode !== 404) throw e;
+    }
+    return this.addService(name, args);
```

Comment:
from a readability perspective, I find this line odd. At a first look, given that this line is outside the try/catch block, we're adding a service regardless of whether the lines above throw or not. In practice, (because of line 550), we're only running `addService` when there is an error. 
I'd propose moving this line to the catch block